src/server/json_family.cc

```diff
@@ -771,8 +772,126 @@ OpResult<vector<OptSizeT>> OpArrInsert(const OpArgs& op_args, string_view key, s
   return vec;
 }
 
+// Returns a numeric vector representing each JSON value first index of the JSON scalar.
+// An index value of -1 represents unfound in the array.
+// JSON scalar has types of string, boolean, null, and number.
+OpResult<vector<OptLong>> OpArrIndex(const OpArgs& op_args, string_view key,
+                                     JsonExpression expression, const json& search_val,
+                                     int start_index, int end_index) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
+  }
+
+  vector<OptLong> vec;
+  auto cb = [&](const string_view& path, const json& val) {
+    if (!val.is_array()) {
+      vec.emplace_back(nullopt);
+      return;
+    }
+
+    if (val.empty()) {
+      vec.emplace_back(-1);
+      return;
+    }
+
+    // Negative value or out-of-range index is handled by rounding the index to the array's start
+    // and end. example: for array size 9 and index -11 the index mapped to index 7.
+    if (start_index < 0) {
+      start_index %= val.size();
+      start_index += val.size();
+    }
+
+    // See the comment above.
+    // A value index of 0 means searching until the end of the array.
+    if (end_index == 0) {
+      end_index = val.size();
+    } else if (end_index < 0) {
+      end_index %= val.size();
+      end_index += val.size();
+    }
+
+    if (start_index > end_index) {
+      vec.emplace_back(-1);
+      return;
+    }
+
+    size_t pos = -1;
+    auto it = next(val.array_range().begin(), start_index);
+    while (it != val.array_range().end()) {
+      if (search_val == *it) {
+        pos = start_index;
+        break;
+      }
+
+      ++it;
+      if (++start_index == end_index) {
+        break;
+      }
+    }
+
+    vec.emplace_back(pos);
+  };
+
+  expression.evaluate(*result, cb);
+  return vec;
+}
+
 }  // namespace
 
+void JsonFamily::ArrIndex(CmdArgList args, ConnectionContext* cntx) {
+  string_view key = ArgS(args, 1);
+  string_view path = ArgS(args, 2);
+
+  error_code ec;
+  JsonExpression expression = jsonpath::make_expression<json>(path, ec);
+
+  if (ec) {
+    VLOG(1) << "Invalid JSONPath syntax: " << ec.message();
+    (*cntx)->SendError(kSyntaxErr);
+    return;
+  }
+
+  optional<json> search_value = ConstructJsonFromString(ArgS(args, 3));
+  if (!search_value) {
+    (*cntx)->SendError(kSyntaxErr);
+    return;
+  }
+
+  if (search_value->is_object() || search_value->is_array()) {
+    (*cntx)->SendError(kWrongTypeErr);
+    return;
+  }
+
+  int start_index = 0;
+  if (args.size() >= 5) {
+    if (!absl::SimpleAtoi(ArgS(args, 4), &start_index)) {
+      VLOG(1) << "Failed to convert the start index to numeric" << ArgS(args, 4);
```

Comment:
why you do not return the error if parsing failed? here and with `end_index` ?

Reply:
actually, I thought about that before, but then I check what happened in Redis Implementation of JSON, and I found they ignore completely those index values and proceed with the search operation if there is a parse error.

Reply:
I do not want to mirror this behavior. You can add a comment saying that it differs from redis but I prefer to reject incorrect arguments similarly to how it's done with other commands.

Reply:
Fixed.